EntranceRando.py

```diff
@@ -0,0 +1,414 @@
+import itertools
+import logging
+import random
+import time
+from collections import deque
+from typing import Callable, Dict, Iterable, List, Tuple, Union, Set, Optional
+
+from BaseClasses import CollectionState, Entrance, Region, EntranceType
+from Options import Accessibility
+from worlds.AutoWorld import World
+
+
+class EntranceRandomizationError(RuntimeError):
+    pass
+
+
+class EntranceLookup:
+    class GroupLookup:
+        _lookup: Dict[str, List[Entrance]]
+
+        def __init__(self):
+            self._lookup = {}
+
+        def __len__(self):
+            return sum(map(len, self._lookup.values()))
+
+        def __bool__(self):
+            return bool(self._lookup)
+
+        def __getitem__(self, item: str) -> List[Entrance]:
+            return self._lookup.get(item, [])
+
+        def __iter__(self):
+            return itertools.chain.from_iterable(self._lookup.values())
+
+        def __str__(self):
+            return str(self._lookup)
+
+        def __repr__(self):
+            return self.__str__()
+
+        def add(self, entrance: Entrance) -> None:
+            self._lookup.setdefault(entrance.er_group, []).append(entrance)
+
+        def remove(self, entrance: Entrance) -> None:
+            group = self._lookup.get(entrance.er_group, [])
```

Comment:
Yeah it would be though the amount of time I spent checking probably indicates this implementation is more clear and definitely does what it should rather than being reliant on the implementation of getitem